1D clustering Yadav.R

```diff
@@ -0,0 +1,282 @@
+  mean_curve_1D <- mean(curve_dnorm_1D)
+  
+  Cscore_1D <- pnorm(spines_clustered_1D, mean_test_1D, std_test_1D) #calculates the probability that given the random sample distribution (curve mean+SD) the total number of spines observed is higher
+  #therefore, the closer to 1, higher probability that a given number has more clustered spines than the random normal distribution and vice versa
+  
+ # add all 1D data to all data 
+  cluster_data_1D <- data.frame()  
+  cluster_data_1D <- rbind(cluster_data_1D, num_clusters_1D)
+  cluster_data_1D <- cbind(cluster_data_1D, spines_clustered_1D)
+  cluster_data_1D <- cbind(cluster_data_1D, spines_not_1D)
+  cluster_data_1D <- cbind(cluster_data_1D, Cscore_1D)
+  cluster_data_1D[is.na(cluster_data_1D)] <- 0
+  colnames(cluster_data_1D) <- c("# of clusters - 1D", "spines clustered - 1D", "spines not clustered - 1D", "Cscore - 1D")
+  
+  
+# 3D clustering analysis  
+  dist_3D <- as.matrix(dist.xyz(data_coords)) # creates distance matrix of spine head coordinates
+  UPGMA_obsv_3D <- IdClusters(dist_3D, method = "UPGMA", cutoff=0.75, showPlot=TRUE) #run cluster analysis with cutoff used in Yadav paper
+  #gives cluster number associated with which spine (i.e. spines 25 and 26 are in cluster 1)
+  UPGMA_obsv_3D$rn <- rownames(UPGMA_obsv_3D)   #adds a column with rownows to keep spine ID, not sure if this step is neccessary
+  cluster_all_3D <- cbind(UPGMA_obsv_3D, data_coords)   #get the coordinates for each spine ID, if kept in row name/number order, will correctly correspond to each spine
+  cluster_all_3D <- cluster_all_3D[order(cluster_all_3D$cluster),] #sort data frame by cluster number
+  cluster_freq_3D <- table(cluster_all_3D$cluster) #create a table counting how many times each cluster Variable occurs (i.e. how many spines in each cluster)
+  cluster_freq_3D <- as.data.frame(cluster_freq_3D)
+  cluster_freq_3D$Freq <- as.numeric(cluster_freq_3D$Freq)
+  cluster_freq_3D$is_clustered <- as.numeric(cluster_freq_3D$Freq >1) # create column where 1 means there is more than one spine in a cluster or 0 if just 1
+  num_clusters_3D <- sum(cluster_freq_3D$is_clustered) #count how many 1s to determine how many clusters (spines > 1) in the segment
+  spines_in_cluster_3D <- cluster_freq_3D %>% group_by(is_clustered) %>% summarise(num_clus_spines_3D = sum(Freq))
+  spines_clustered_3D <- list()
+  spines_clustered_3D <- rbind(spines_clustered_3D, spines_in_cluster_3D[2,2])
+  spines_clustered_3D <- as.matrix(spines_clustered_3D) #conerts num of spines clustered to matrix
+  spines_clustered_3D <- as.numeric(spines_clustered_3D) #converts to numerical form
+  
+  spines_not_3D <- as.numeric(total_spines - spines_clustered_3D)
+  
+  test_spines_clustered_all_3D <- data.frame()
+  test_num_clusters_all_3D <- data.frame()
+
+  
+# 3D random spines for loop
+  for(j in 1:100){
+    test_data_X <- data.frame(sample(df$X), df$Y, df$Z) # randomize the X's, Y's, and Z's to make a "biologically plausible" dataframe.
```

Comment:
I think only the first line and last line in this block are necessary:
```
test_data <- data.frame(sample(df$X), sample(df$Y), sample(df$Z))
test_dist_3D <- as.matrix(dist(test_data)) # creates distance matrix for random sample
```
But I think I see your thought process to constrain how "random" the datapoints are, so I'll just pin this for further conversation.

Reply:
I think I have an idea for how to more accurately constrain the datapoints, since the random points still seem to be a little too random. (we **should** expect clustering on most dendritic branches, we just want to see changes in the degree).

I'm thinking that we need to set up some sort of selection criteria for the random coordinates.
The dendrites and spines essentially live in a known cylindrical space (length being the length of the dendrite which we have 3D coordinates to regenerate, and the width/radius being the max spine length--currently when I map the random 3D coordinates it doesn't really **look** like spines along a dendritic length.

So, if we make criteria that the random 3D coordinates must fall within the known limits, we may be able to have better randomization. I'm going to try thinking about that this week, but let me know what you think.

Reply:
That sounds like a good idea, but let me think through it.
I'll make up some numbers (that may not be biologically plausible).

if we have a 50um dendrite segment with 150 spine heads, then we have 150 x-coordinates, 150 y-coordinates and 150 z-coordinates. 
If we randomized using the data available, we would have 150*150*150 or 150^3 or 3375000 possible combinations of 150 coordinates.

If we generate a list of possible coordinates within a cylinder incrementing by 0.01, this list should include the 3375000 coordinates produced by the previous method (since all those xyz-coordinates were actually observed and recorded) and several billion others.
napkin math (if I assume a radius of 5um):
volume of cylinder: 50*(pi*5^2) ~ 3927um
increments of 0.01: 3927 / 0.01 = 392700 coordinates
all combinations of selecting 150 points: 392700 choose 150 ~ 2187165081225137310382017233099995046337100876544142411431297970118841785272017984540668630264009105699815020447043722621456400930097774157111152285573435447744179823216702996167899049246262375503173540118312468464288508173005027094694777928571049706316046267083117726289897816726177377941620617732303764895199090971241059657147091983664496692488541826387269341802889197146363386214674094117662445818311269533829858635004649504925878489693364264340423726784051623161387140226054951726805351551643633068009282354287466945089405647311894109982832344478823923883759838211381115148 possible combinations of 150 coordinates

(I did not use a napkin).

From this example, randomizing the observed data constrains the space where spine heads "can" be by many orders of magnitude. I think the hop from 1D space to 3D space adds more dimensions for data to move and thus completely random data are less likely to cluster, and I don't know if there is an easy way to get around that, besides adding some conditionals that say spines that are too far from the next closest spine should be reselected.

Using the observed data should look closer to the biological reality (but probably not anywhere close to perfect), and the resulting coordinates should be within the cylinder of interest since we are using the observed data to generate "random" data points. Generating all possible datapoints will create a bunch of new places the spine heads can be, but I cannot think of a reason why they would be more clustered since we are still dealing with the problem of 3 dimensions versus 1 dimension (it's easier to cluster in 1 dimension than 3).

Reply:
Since the coordinates do not *look* right, do we need to include information about SOMA-DISTANCE in addition to the x-y-z coordinates? perhaps I'm still confused on what each variable represents in space.